# Project 4: Data Analysis and Visualization with Pandas and Streamlit

# --- 1️⃣ Import and load data ---

In [1]:
import pandas as pd
df = pd.read_csv("/Users/nili/sp500/constituents-financials.csv")
df.head()

Name                  Sector   Price  Price/Earnings  \
0           3M Company             Industrials  222.89           24.31   
1      A.O. Smith Corp             Industrials   60.24           27.76   
2  Abbott Laboratories             Health Care   56.27           22.51   
3          AbbVie Inc.             Health Care  108.48           19.41   
4        Accenture plc  Information Technology  150.51           25.47   

   Dividend Yield  Earnings/Share  52 Week Low  52 Week High    Market Cap  \
0        2.332862            7.92       259.77       175.490  1.390000e+11   
1        1.147959            1.70        68.39        48.925  1.078342e+10   
2        1.908982            0.26        64.60        42.280  1.020000e+11   
3        2.499560            3.29       125.86        60.050  1.810000e+11   
4        1.714470            5.44       162.60       114.820  9.876586e+10   

         EBITDA  Price/Sales  Price/Book  
0  9.048000e+09     4.390271       11.34  
1  6.010000e+08     3.575483        6.35  
2  5.744000e+09     3.740480        3.19  
3  1.031000e+10     6.291571       26.14  
4  5.643228e+09     2.604117       10.62

In [2]:
df

Name                  Sector   Price  Price/Earnings  \
0                3M Company             Industrials  222.89           24.31   
1           A.O. Smith Corp             Industrials   60.24           27.76   
2       Abbott Laboratories             Health Care   56.27           22.51   
3               AbbVie Inc.             Health Care  108.48           19.41   
4             Accenture plc  Information Technology  150.51           25.47   
..                      ...                     ...     ...             ...   
500              Xylem Inc.             Industrials   70.24           30.94   
501         Yum! Brands Inc  Consumer Discretionary   76.30           27.25   
502  Zimmer Biomet Holdings             Health Care  115.53           14.32   
503           Zions Bancorp              Financials   50.71           17.73   
504                  Zoetis             Health Care   71.51           32.80   

     Dividend Yield  Earnings/Share  52 Week Low  52 Week High    Market Cap  \
0          2.332862            7.92       259.77       175.490  1.390000e+11   
1          1.147959            1.70        68.39        48.925  1.078342e+10   
2          1.908982            0.26        64.60        42.280  1.020000e+11   
3          2.499560            3.29       125.86        60.050  1.810000e+11   
4          1.714470            5.44       162.60       114.820  9.876586e+10   
..              ...             ...          ...           ...           ...   
500        1.170079            1.83        76.81        46.860  1.291502e+10   
501        1.797080            4.07        86.93        62.850  2.700330e+10   
502        0.794834            9.01       133.49       108.170  2.445470e+10   
503        1.480933            2.60        55.61        38.430  1.067068e+10   
504        0.682372            1.65        80.13        52.000  3.599111e+10   

           EBITDA  Price/Sales  Price/Book  
0    9.048000e+09     4.390271       11.34  
1    6.010000e+08     3.575483        6.35  
2    5.744000e+09     3.740480        3.19  
3    1.031000e+10     6.291571       26.14  
4    5.643228e+09     2.604117       10.62  
..            ...          ...         ...  
500  7.220000e+08     2.726209        5.31  
501  2.289000e+09     6.313636      212.08  
502  2.007400e+09     3.164895        2.39  
503  0.000000e+00     3.794579        1.42  
504  1.734000e+09     9.280896       18.09  

[505 rows x 12 columns]

# --- 2️⃣ Check info and missing data ---

In [3]:
print("\n=== Data Info ===")
df.info()


=== Data Info ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 505 entries, 0 to 504
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Name            505 non-null    object 
 1   Sector          505 non-null    object 
 2   Price           505 non-null    float64
 3   Price/Earnings  503 non-null    float64
 4   Dividend Yield  505 non-null    float64
 5   Earnings/Share  505 non-null    float64
 6   52 Week Low     505 non-null    float64
 7   52 Week High    505 non-null    float64
 8   Market Cap      505 non-null    float64
 9   EBITDA          505 non-null    float64
 10  Price/Sales     505 non-null    float64
 11  Price/Book      497 non-null    float64
dtypes: float64(10), object(2)
memory usage: 47.5+ KB


In [4]:
print("\n=== Missing Values ===")
print(df.isna().sum())


=== Missing Values ===
Name              0
Sector            0
Price             0
Price/Earnings    2
Dividend Yield    0
Earnings/Share    0
52 Week Low       0
52 Week High      0
Market Cap        0
EBITDA            0
Price/Sales       0
Price/Book        8
dtype: int64


# Fill missing numeric columns with mean

In [5]:
df["Price/Earnings"] = df["Price/Earnings"].fillna(df["Price/Earnings"].mean())
df["Price/Book"] = df["Price/Book"].fillna(df["Price/Book"].mean())
df.isna().sum()

Name              0
Sector            0
Price             0
Price/Earnings    0
Dividend Yield    0
Earnings/Share    0
52 Week Low       0
52 Week High      0
Market Cap        0
EBITDA            0
Price/Sales       0
Price/Book        0
dtype: int64

# --- 3️⃣ Basic Statistics ---

In [6]:
print("\n=== Summary Statistics ===")
print(df.describe())


=== Summary Statistics ===
             Price  Price/Earnings  Dividend Yield  Earnings/Share  \
count   505.000000      505.000000      505.000000      505.000000   
mean    103.830634       24.808390        1.895953        3.753743   
std     134.427636       41.159172        1.537214        5.689036   
min       2.820000     -251.530000        0.000000      -28.010000   
25%      46.250000       15.350000        0.794834        1.490000   
50%      73.920000       19.540000        1.769255        2.890000   
75%     116.540000       25.740000        2.781114        5.140000   
max    1806.060000      520.150000       12.661196       44.090000   

       52 Week Low  52 Week High    Market Cap        EBITDA  Price/Sales  \
count   505.000000    505.000000  5.050000e+02  5.050000e+02   505.000000   
mean    122.623832     83.536616  4.923784e+10  3.590328e+09     3.941705   
std     155.362140    105.725473  9.006020e+10  6.840544e+09     3.460110   
min       6.590000      2.800000 

# --- 4️⃣ Example analyses ---

In [7]:
tech = df[df["Sector"] == "Information Technology"]
print(tech)

                           Name                  Sector   Price  \
4                 Accenture plc  Information Technology  150.51   
5           Activision Blizzard  Information Technology   65.83   
7             Adobe Systems Inc  Information Technology  185.16   
9    Advanced Micro Devices Inc  Information Technology   11.22   
16      Akamai Technologies Inc  Information Technology   62.49   
..                          ...                     ...     ...   
477                   Visa Inc.  Information Technology  113.86   
487             Western Digital  Information Technology   80.09   
488            Western Union Co  Information Technology   18.81   
497                 Xerox Corp.  Information Technology   29.80   
498                  Xilinx Inc  Information Technology   62.82   

     Price/Earnings  Dividend Yield  Earnings/Share  52 Week Low  \
4             25.47        1.714470            5.44      162.600   
5             31.80        0.431903            1.28       7

## 4.1 Top 10 companies by EBITDA

In [8]:
top10_ebitda = df.sort_values('EBITDA', ascending=False).head(10)
print(top10_ebitda)

                       Name                      Sector    Price  \
51               Apple Inc.      Information Technology   155.15   
58                 AT&T Inc  Telecommunication Services    35.57   
474  Verizon Communications  Telecommunication Services    49.04   
311         Microsoft Corp.      Information Technology    85.01   
185       Exxon Mobil Corp.                      Energy    76.07   
27     Alphabet Inc Class A      Information Technology  1007.71   
28     Alphabet Inc Class C      Information Technology  1001.52   
480         Wal-Mart Stores            Consumer Staples   100.02   
105           Chevron Corp.                      Energy   112.30   
122           Comcast Corp.      Consumer Discretionary    38.19   

     Price/Earnings  Dividend Yield  Earnings/Share  52 Week Low  \
51            16.86        1.579541            9.20       180.10   
58            12.14        5.415651            4.76        42.70   
474           13.08        4.626544            

## 4.2 Average Price

In [9]:
average_price = df['Price'].mean()
print(f"Average stock price: {average_price:.2f}")

Average stock price: 103.83


## 4.3 High price stocks

In [10]:
High_price_stock = df[df["Price"] >= 1000]
print(High_price_stock)


                     Name                  Sector    Price  Price/Earnings  \
27   Alphabet Inc Class A  Information Technology  1007.71           31.48   
28   Alphabet Inc Class C  Information Technology  1001.52           40.29   
30         Amazon.com Inc  Consumer Discretionary  1350.50          296.16   
372     Priceline.com Inc  Consumer Discretionary  1806.06           24.26   

     Dividend Yield  Earnings/Share  52 Week Low  52 Week High    Market Cap  \
27              0.0           22.27      1198.00      824.3000  7.340000e+11   
28              0.0           22.27      1186.89      803.1903  7.290000e+11   
30              0.0            6.16      1498.00      812.5000  6.860000e+11   
372             0.0           42.66      2067.99     1589.0000  9.181745e+10   

           EBITDA  Price/Sales  Price/Book  
27   3.421700e+10     6.801692        4.70  
28   3.271400e+10     6.772653        4.67  
30   1.613200e+10     3.927053       24.28  
372  4.803487e+09     9.1765

# --- 5️⃣ Advanced analysis: EPS vs P/E relationship ---

Companies whose Earnings per Share (EPS) are above the median (2.89)
and whose Price-to-Earnings ratio (P/E) is below the median (19.54).
These companies have stronger profitability but reasonable valuation.

In [11]:
eps_median = df["Earnings/Share"].median()
pe_median = df["Price/Earnings"].median()

strong_stocks = df[
    (df["Earnings/Share"] > eps_median) &
    (df["Price/Earnings"] < pe_median)
]

print("\n=== Strong Companies (High EPS & Low P/E) ===")
print(strong_stocks[["Sector", "Earnings/Share", "Price/Earnings"]].head(20))


=== Strong Companies (High EPS & Low P/E) ===
                        Sector  Earnings/Share  Price/Earnings
3                  Health Care            3.29           19.41
6                  Industrials            7.43           18.22
11                 Health Care            5.75           18.11
12                  Financials           12.07           12.24
13                  Financials           11.01           12.24
17                 Industrials            8.28            9.66
23                 Health Care           38.35           10.65
24      Information Technology           14.13           13.02
26                  Financials            4.68           13.26
29            Consumer Staples            5.31           18.89
32                 Industrials            3.91            9.92
34                  Financials            2.90           15.00
38                  Financials            9.44           12.41
45                      Energy            6.13           18.78
47      

In [12]:
print("Strong companies:", len(strong_stocks))

Strong companies: 135


# --- 6️⃣ Grouping by Sector ---

In [13]:
print("\n=== Strong Companies per Sector ===")
print(strong_stocks["Sector"].value_counts())


=== Strong Companies per Sector ===
Sector
Financials                    30
Consumer Discretionary        26
Health Care                   18
Industrials                   17
Information Technology         9
Real Estate                    9
Utilities                      8
Consumer Staples               7
Materials                      6
Energy                         3
Telecommunication Services     2
Name: count, dtype: int64


In [14]:
grouped = strong_stocks.groupby("Sector").size()
print(grouped)

Sector
Consumer Discretionary        26
Consumer Staples               7
Energy                         3
Financials                    30
Health Care                   18
Industrials                   17
Information Technology         9
Materials                      6
Real Estate                    9
Telecommunication Services     2
Utilities                      8
dtype: int64


In [15]:
sector_summary = strong_stocks.groupby("Sector")[["Earnings/Share", "Price/Earnings"]].mean()
print("\n=== Average EPS & P/E per Sector ===")
print(sector_summary)


=== Average EPS & P/E per Sector ===
                            Earnings/Share  Price/Earnings
Sector                                                    
Consumer Discretionary            6.678077       14.648077
Consumer Staples                  5.652857       15.641429
Energy                            7.406667       17.860000
Financials                        5.895000       14.597333
Health Care                       9.330000       14.797222
Industrials                       7.064118       15.994118
Information Technology            6.832222       14.323333
Materials                         7.498333       15.418333
Real Estate                       4.755556       17.105556
Telecommunication Services        6.060000       12.610000
Utilities                         4.065000       16.232500


# --- 7️⃣ Find the best-performing sector ---

In [16]:
best_sector = sector_summary["Earnings/Share"].idxmax()
print("Sector with highest average EPS:", best_sector)

Sector with highest average EPS: Health Care


In [17]:
sector_summary = (
    strong_stocks
    .groupby("Sector")[["Earnings/Share"]]
    .mean()
    .sort_values("Earnings/Share", ascending=False)
)

best_sector = sector_summary.head(10)
print("Best sector for strong stocks:", best_sector)

Best sector for strong stocks:                             Earnings/Share
Sector                                    
Health Care                       9.330000
Materials                         7.498333
Energy                            7.406667
Industrials                       7.064118
Information Technology            6.832222
Consumer Discretionary            6.678077
Telecommunication Services        6.060000
Financials                        5.895000
Consumer Staples                  5.652857
Real Estate                       4.755556



# --- 8️⃣ Save results to CSV files ---

In [18]:
df.to_csv("sp500_clean.csv", index=False)
strong_stocks.to_csv("strong_stocks.csv", index=False)
sector_summary.to_csv("sector_summary.csv")

print("\n✅ All results saved: sp500_clean.csv, strong_stocks.csv, sector_summary.csv")


✅ All results saved: sp500_clean.csv, strong_stocks.csv, sector_summary.csv


In this project, I analyzed the S&P 500 company financials dataset using pandas.
After cleaning missing data and exploring key statistics, I compared companies by Earnings per Share (EPS) and Price-to-Earnings (P/E) ratio.
Companies with above-median EPS and below-median P/E were identified as strong stocks — meaning they are more profitable and reasonably valued.

By grouping these strong companies by sector, I found that the Health Care sector had the highest average EPS (≈ 9.33), making it the best-performing sector overall.
Finally, the cleaned data and summarized results were saved into new CSV files for future analysis.

In [19]:
%%writefile app.py
import streamlit as st
import pandas as pd

st.title("📊 S&P 500 Financial Analysis")



Overwriting app.py
